indra/newview/llfloateremojipicker.cpp

```diff
@@ -313,299 +346,492 @@ void LLFloaterEmojiPicker::dirtyRect()
         }
+    }
 
-        for (const LLEmojiDescriptor* descr : emojis)
+    if (mEmojiGrid->getPanelList().empty())
+    {
+        showPreview(false);
+        mFocusedIconRow = -1;
+        mFocusedIconCol = -1;
+        if (!mHint.empty())
         {
-            if (sFilterPattern.empty() || matchesPattern(descr))
-            {
-                // Place a category title if needed
-                if (showDivider)
-                {
-                    LLEmojiGridDivider* div = new LLEmojiGridDivider(row_panel_params, category);
-                    mEmojiGrid->addPanel(div, true);
-                    showDivider = false;
-                }
-
-                // Place a new row each (maxIcons) icons
-                if (!(iconIndex % maxIcons))
-                {
-                    if (maxRows && ++rowCount > maxRows)
-                        break;
-                    row = new LLEmojiGridRow(row_panel_params, row_list_params);
-                    mEmojiGrid->addPanel(row, true);
-                }
-
-                // Place a new icon to the current row
-                LLEmojiGridIcon* icon = new LLEmojiGridIcon(icon_params, descr, mixedFolder ? LLStringUtil::capitalize(descr->Category) : category);
-                icon->setMouseEnterCallback([this](LLUICtrl* ctrl, const LLSD&) { onEmojiMouseEnter(ctrl); });
-                icon->setMouseLeaveCallback([this](LLUICtrl* ctrl, const LLSD&) { onEmojiMouseLeave(ctrl); });
-                icon->setMouseDownCallback([this](LLUICtrl* ctrl, S32, S32, MASK) { onEmojiMouseDown(ctrl); });
-                icon->setMouseUpCallback([this](LLUICtrl* ctrl, S32, S32, MASK) { onEmojiMouseUp(ctrl); });
-                icon->setBackgroundColor(bgColor);
-                icon->setBackgroundOpaque(1);
-                icon->setRect(icon_rect);
-                row->mList->addPanel(icon, true);
-
-                iconIndex++;
-            }
+            hideFloater();
         }
-    };
+    }
+    else
+    {
+        showPreview(true);
+        mFocusedIconRow = 0;
+        mFocusedIconCol = 0;
+        moveFocusedIconNext();
+    }
+}
 
-    const std::vector<LLEmojiGroup>& groups = LLEmojiDictionary::instance().getGroups();
-    const LLEmojiDictionary::emoji2descr_map_t& emoji2descr = LLEmojiDictionary::instance().getEmoji2Descr();
-    const LLEmojiDictionary::cat2descrs_map_t& category2Descr = LLEmojiDictionary::instance().getCategory2Descrs();
-    if (!sSelectedGroupIndex)
+void LLFloaterEmojiPicker::fillEmojisCategory(const std::vector<LLEmojiSearchResult>& emojis,
+    const std::string& category, const LLPanel::Params& row_panel_params, const LLUICtrl::Params& row_list_params,
+    const LLPanel::Params& icon_params, const LLRect& icon_rect, S32 max_icons, const LLColor4& bg)
+{
+    // Place the category title
+    std::string title =
+        category == RECENTLY_USED_CATEGORY ? getString("title_for_recently_used") :
```

Comment:
Does this work correctly on other window's locales (like german)?

Reply:
I meant the RECENTLY_USED_CATEGORY

Reply:
There are 2 special categories which are contained not in dictionary, they are processed special ways